### PR TITLE
feat: recover from mls-stale-message WPB-538

### PR DESCRIPTION
**Pull request by David-Henner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-538" title="WPB-538" target="_blank">StoryWPB-538</a>  [iOS] MLS Client recover from "not able to encrypt a message"
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

When attempting to send a message in a conversation, a client can discover that the conversation is out of sync (the local epoch is behind the remote epoch) by receiving a 409 `mls-stale-message` response from the backend. 
When this happens we should repair the MLS group and inform the user that they may have missed messages. Then we can retry to encrypt and send the message for the correct epoch.

## Changes

- Handling `mis-stale-message` when sending commits (in `MLSActionExecutor`)
- Handling `mls-stale-message` when sending MLS messages (in `MLSMessageSync`)
- Added functionality to `SubconversationGroupIDRepository` to find parent group ID from a subgroup ID
- Fixed concurrency issue due to accessing the `conversationsBeingRepaired` set from multiple threads. Now using an actor to ensure thread safety by serialising access to its properties
- Added a developer action to ignore incoming events
- Added unit tests

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Object Syncs/Helpers/MLSMessageSync.swift [hunk 1]
```diff
@@ -62,28 +66,58 @@ class MLSMessageSync<Message: MLSMessage>: NSObject, ZMContextChangeTrackerSourc
...
+                    let error = SendMLSMessageAction.Failure(from: response),
+                    error == .mlsStaleMessage
+                else {
+                    return completion(result, response)
+                }
+
+                Task {
+                    WireLogger.mls.info("got stale message error when sending message in group (\(groupID.safeForLoggingDescription)). rejoining group and trying again...")
+                    await mlsService.fetchAndRepairGroup(with: groupID)
+                    self.synchronize(entity: message, in: self.context, completion: completion)
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Object Syncs/Helpers/MLSMessageSync.swift:93]: If we synchronize again, perhaps we need to post the new request available notification once more.

File: wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift [hunk 2]
```diff
@@ -65,4 +68,20 @@ public final class SubconversationGroupIDRepository: SubconversationGroupIDRepos
...
 
+    // MARK: - Finding subgroup
+
+    public func findSubgroupTypeAndParentID(
+        for targetGroupID: MLSGroupID
+    ) -> (parentID: MLSGroupID, type: SubgroupType)? {
+
+        for (parentID, subgroups) in storage {
+            if let subgroup = subgroups.first(where: {
+                $1 == targetGroupID
```
Comment (netbe) [inline wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift]: what's $1 here?

**Review by netbe — APPROVED** [review]
just one question

File: wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift (hunk 2 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift]: `$1` is the subgroup ID

Comment (netbe) [reply wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift]: what's $0? maybe we use explicit name, or is it me just unfamiliar with the subgroupType

Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift]: updated [4887f50](https://github.com/wireapp/wire-ios/pull/598/commits/4887f50436ba14a75839a99c7447b530469a9c68)

File: wire-ios-request-strategy/Sources/Object Syncs/Helpers/MLSMessageSync.swift (hunk 1 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/Object Syncs/Helpers/MLSMessageSync.swift:93]: done [4887f50](https://github.com/wireapp/wire-ios/pull/598/commits/4887f50436ba14a75839a99c7447b530469a9c68)

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**